<a href="https://colab.research.google.com/github/bbamknp2018/MSC_DPDM_2025/blob/main/Hw8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [30]:
import pandas as pd
import os

In [31]:
path = '/content/'

In [32]:
data1= pd.read_csv(os.path.join(path,'2025-01-city-of-london-stop-and-search.csv'),encoding='latin-1') #encoding= 'latin-1' bc. มีภาษาที่เป็นจุด ๆ บนหัวทำให้ข้อมมูลเละ
data1

Type                       Date  \
0    Person and Vehicle search  2025-01-01T19:44:54+00:00   
1    Person and Vehicle search  2025-01-01T19:53:21+00:00   
2    Person and Vehicle search  2025-01-01T20:19:35+00:00   
3    Person and Vehicle search  2025-01-01T23:12:51+00:00   
4                Person search  2025-01-02T03:19:44+00:00   
..                         ...                        ...   
149  Person and Vehicle search  2025-01-31T11:28:02+00:00   
150              Person search  2025-01-31T13:26:41+00:00   
151              Person search  2025-01-31T13:40:18+00:00   
152              Person search  2025-01-31T16:01:47+00:00   
153              Person search  2025-01-31T22:29:53+00:00   

     Part of a policing operation  Policing operation   Latitude  Longitude  \
0                             NaN                 NaN        NaN        NaN   
1                             NaN                 NaN        NaN        NaN   
2                             NaN                 NaN        NaN        NaN   
3                             NaN                 NaN        NaN        NaN   
4                             NaN                 NaN  51.510042  -0.120439   
..                            ...                 ...        ...        ...   
149                           NaN                 NaN  51.513666  -0.078278   
150                           NaN                 NaN  51.512105  -0.085089   
151                           NaN                 NaN  51.512105  -0.085089   
152                           NaN                 NaN  51.512658  -0.092531   
153                           NaN                 NaN  51.517326  -0.079883   

     Gender Age range                             Self-defined ethnicity  \
0      Male     18-24                  Asian/Asian British - Bangladeshi   
1      Male     18-24                  Asian/Asian British - Bangladeshi   
2      Male   over 34                    Other ethnic group - Not stated   
3      Male     25-34    Black/African/Caribbean/Black British - African   
4      Male     18-24                    Other ethnic group - Not stated   
..      ...       ...                                                ...   
149    Male   over 34        Other ethnic group - Any other ethnic group   
150    Male   over 34  Black/African/Caribbean/Black British - Any ot...   
151  Female       NaN                    Other ethnic group - Not stated   
152    Male   over 34                 White - Any other White background   
153    Male     18-24  Black/African/Caribbean/Black British - Any ot...   

    Officer-defined ethnicity  \
0                       Asian   
1                       White   
2                       Asian   
3                       Black   
4                       Asian   
..                        ...   
149                     Other   
150                     Black   
151                     Black   
152                     White   
153                     Black   

                                           Legislation  \
0                Misuse of Drugs Act 1971 (section 23)   
1                Misuse of Drugs Act 1971 (section 23)   
2                Misuse of Drugs Act 1971 (section 23)   
3                Misuse of Drugs Act 1971 (section 23)   
4                Misuse of Drugs Act 1971 (section 23)   
..                                                 ...   
149              Misuse of Drugs Act 1971 (section 23)   
150  Police and Criminal Evidence Act 1984 (section 1)   
151  Police and Criminal Evidence Act 1984 (section 1)   
152  Police and Criminal Evidence Act 1984 (section 1)   
153              Misuse of Drugs Act 1971 (section 23)   

             Object of search                       Outcome  \
0            Controlled drugs  A no further action disposal   
1            Controlled drugs  A no further action disposal   
2            Controlled drugs  A no further action disposal   
3            Controlled drugs  A no further action disposal   
4            Controlle

In [7]:
data2= pd.read_csv(os.path.join(path,'2025-02-city-of-london-stop-and-search.csv'),encoding='latin-1') #encoding= 'latin-1' bc. มีภาษาที่เป็นจุด ๆ บนหัวทำให้ข้อมมูลเละ
data2

Type                       Date  \
0                Person search  2025-02-01T04:35:33+00:00   
1                Person search  2025-02-01T14:24:45+00:00   
2                Person search  2025-02-01T21:28:25+00:00   
3                Person search  2025-02-01T21:35:37+00:00   
4    Person and Vehicle search  2025-02-02T01:02:10+00:00   
..                         ...                        ...   
120              Person search  2025-02-28T18:45:46+00:00   
121  Person and Vehicle search  2025-02-28T20:03:44+00:00   
122              Person search  2025-02-28T21:25:07+00:00   
123              Person search  2025-02-28T21:30:36+00:00   
124              Person search  2025-02-28T22:12:07+00:00   

     Part of a policing operation  Policing operation   Latitude  Longitude  \
0                             NaN                 NaN        NaN        NaN   
1                             NaN                 NaN  51.514267  -0.079838   
2                             NaN                 NaN  51.511406  -0.075866   
3                             NaN                 NaN  51.511965  -0.075453   
4                             NaN                 NaN  51.509797  -0.083600   
..                            ...                 ...        ...        ...   
120                           NaN                 NaN  51.510725  -0.084844   
121                           NaN                 NaN  51.513100  -0.072542   
122                           NaN                 NaN  51.514626  -0.097291   
123                           NaN                 NaN  51.514626  -0.097291   
124                           NaN                 NaN  51.510725  -0.084844   

    Gender Age range                             Self-defined ethnicity  \
0     Male     25-34  White - English/Welsh/Scottish/Northern Irish/...   
1     Male   over 34                 White - Any other White background   
2     Male     18-24  White - English/Welsh/Scottish/Northern Irish/...   
3     Male     18-24  White - English/Welsh/Scottish/Northern Irish/...   
4     Male   over 34   Asian/Asian British - Any other Asian background   
..     ...       ...                                                ...   
120   Male     10-17                    Other ethnic group - Not stated   
121   Male     18-24                    Other ethnic group - Not stated   
122   Male     25-34                 White - Any other White background   
123   Male     18-24                 White - Any other White background   
124   Male     18-24                 White - Any other White background   

    Officer-defined ethnicity  \
0                       White   
1                       White   
2                       White   
3                       White   
4                       Asian   
..                        ...   
120                     White   
121                     Asian   
122                     White   
123                     White   
124                     White   

                                           Legislation  \
0                Misuse of Drugs Act 1971 (section 23)   
1    Police and Criminal Evidence Act 1984 (section 1)   
2                Misuse of Drugs Act 1971 (section 23)   
3                Misuse of Drugs Act 1971 (section 23)   
4                Misuse of Drugs Act 1971 (section 23)   
..                                                 ...   
120  Police and Criminal Evidence Act 1984 (section 1)   
121              Misuse of Drugs Act 1971 (section 23)   
122              Misuse of Drugs Act 1971 (section 23)   
123              Misuse of Drugs Act 1971 (section 23)   
124  Police and Criminal Evidence Act 1984 (section 1)   

                       Object of search                       Outcome  \
0                      Controlled drugs                        Arrest   
1              Article for use in theft  A no further action disposal   
2                      Controlled drugs          Community resolution   
3                      Controlled drugs          Communit

In [8]:
data3= pd.read_csv(os.path.join(path,'2025-03-city-of-london-stop-and-search.csv'),encoding='latin-1') #encoding= 'latin-1' bc. มีภาษาที่เป็นจุด ๆ บนหัวทำให้ข้อมมูลเละ
data3.head()

Type                       Date  \
0              Person search  2025-03-01T21:45:31+00:00   
1              Person search  2025-03-01T23:30:44+00:00   
2  Person and Vehicle search  2025-03-02T15:11:20+00:00   
3              Person search  2025-03-02T22:05:02+00:00   
4  Person and Vehicle search  2025-03-03T18:29:41+00:00   

   Part of a policing operation  Policing operation   Latitude  Longitude  \
0                           NaN                 NaN  51.517785  -0.079907   
1                           NaN                 NaN  51.516869  -0.081113   
2                           NaN                 NaN  51.529250  -0.088709   
3                           NaN                 NaN  51.513736  -0.095224   
4                           NaN                 NaN  51.514202  -0.108450   

  Gender Age range                             Self-defined ethnicity  \
0   Male       NaN  White - English/Welsh/Scottish/Northern Irish/...   
1   Male   over 34                    Other ethnic group - Not stated   
2   Male     25-34  Black/African/Caribbean/Black British - Any ot...   
3   Male     18-24                    Other ethnic group - Not stated   
4   Male     25-34                 White - Any other White background   

  Officer-defined ethnicity  \
0                     White   
1                     White   
2                     Black   
3                     Other   
4                     White   

                                         Legislation  \
0  Police and Criminal Evidence Act 1984 (section 1)   
1              Misuse of Drugs Act 1971 (section 23)   
2                     Firearms Act 1968 (section 47)   
3  Police and Criminal Evidence Act 1984 (section 1)   
4              Misuse of Drugs Act 1971 (section 23)   

           Object of search                       Outcome  \
0         Offensive weapons  A no further action disposal   
1          Controlled drugs  A no further action disposal   
2                  Firearms  A no further action disposal   
3  Article for use in theft                        Arrest   
4          Controlled drugs  A no further action disposal   

   Outcome linked to object of search  \
0                               False   
1                               False   
2                               False   
3                                True   
4                               False   

   Removal of more than just outer clothing  
0                                     False  
1                                     False  
2                                     False  
3                                     False  
4                                     False

In [9]:
data4= pd.read_csv(os.path.join(path,'2025-04-city-of-london-stop-and-search.csv'),encoding='latin-1') #encoding= 'latin-1' bc. มีภาษาที่เป็นจุด ๆ บนหัวทำให้ข้อมมูลเละ
data4

Type                       Date  \
0                Person search  2025-03-31T23:56:17+00:00   
1    Person and Vehicle search  2025-04-01T00:09:14+00:00   
2                Person search  2025-04-01T00:16:09+00:00   
3    Person and Vehicle search  2025-04-01T00:55:24+00:00   
4                Person search  2025-04-01T00:58:50+00:00   
..                         ...                        ...   
150              Person search  2025-04-29T11:12:40+00:00   
151              Person search  2025-04-30T07:51:04+00:00   
152  Person and Vehicle search  2025-04-30T15:24:47+00:00   
153              Person search  2025-04-30T19:00:48+00:00   
154  Person and Vehicle search  2025-04-30T20:50:32+00:00   

     Part of a policing operation  Policing operation   Latitude  Longitude  \
0                             NaN                 NaN  51.526381  -0.078808   
1                             NaN                 NaN  51.508588  -0.074039   
2                             NaN                 NaN  51.508588  -0.074039   
3                             NaN                 NaN  51.517102  -0.088785   
4                             NaN                 NaN  51.517102  -0.088785   
..                            ...                 ...        ...        ...   
150                           NaN                 NaN  51.518692  -0.081512   
151                           NaN                 NaN  51.521127  -0.086858   
152                           NaN                 NaN  51.514975  -0.088456   
153                           NaN                 NaN  51.509967  -0.078059   
154                           NaN                 NaN  51.517102  -0.088785   

     Gender Age range                           Self-defined ethnicity  \
0    Female     18-24  Black/African/Caribbean/Black British - African   
1    Female     25-34                  Other ethnic group - Not stated   
2    Female       NaN                  Other ethnic group - Not stated   
3      Male     25-34                  Other ethnic group - Not stated   
4      Male       NaN                  Other ethnic group - Not stated   
..      ...       ...                                              ...   
150    Male       NaN                  Other ethnic group - Not stated   
151    Male     18-24  Black/African/Caribbean/Black British - African   
152    Male     18-24   Mixed/Multiple ethnic groups - White and Asian   
153    Male       NaN                  Other ethnic group - Not stated   
154    Male     18-24                  Other ethnic group - Not stated   

    Officer-defined ethnicity  \
0                       Other   
1                       White   
2                       White   
3                       Other   
4                       Other   
..                        ...   
150                     White   
151                     Black   
152                     Asian   
153                     Black   
154                     White   

                                           Legislation   Object of search  \
0    Police and Criminal Evidence Act 1984 (section 1)  Offensive weapons   
1                Misuse of Drugs Act 1971 (section 23)   Controlled drugs   
2                Misuse of Drugs Act 1971 (section 23)   Controlled drugs   
3                Misuse of Drugs Act 1971 (section 23)   Controlled drugs   
4                Misuse of Drugs Act 1971 (section 23)   Controlled drugs   
..                                                 ...                ...   
150  Police and Criminal Evidence Act 1984 (section 1)       Stolen goods   
151              Misuse of Drugs Act 1971 (section 23)   Controlled drugs   
152              Misuse of Drugs Act 1971 (section 23)   Controlled drugs   
153              Misuse of Drugs Act 1971 (section 23)   Controlled drugs   
154              Misuse of Drugs Act 1971 (section 23)   Controlled drugs   

                          Outcome  Outcome linked to object of search  \
0    A no further action disposal                           

In [10]:
data5= pd.read_csv(os.path.join(path,'2025-05-city-of-london-stop-and-search.csv'),encoding='latin-1') #encoding= 'latin-1' bc. มีภาษาที่เป็นจุด ๆ บนหัวทำให้ข้อมมูลเละ
data5

Type                       Date  Part of a policing operation  \
0    Person search  2025-05-01T15:15:52+00:00                           NaN   
1    Person search  2025-05-01T18:54:29+00:00                           NaN   
2    Person search  2025-05-01T20:41:52+00:00                           NaN   
3    Person search  2025-05-01T20:46:35+00:00                           NaN   
4    Person search  2025-05-01T21:28:28+00:00                           NaN   
..             ...                        ...                           ...   
145  Person search  2025-05-31T04:25:52+00:00                           NaN   
146  Person search  2025-05-31T05:38:03+00:00                           NaN   
147  Person search  2025-05-31T05:47:40+00:00                           NaN   
148  Person search  2025-05-31T12:51:55+00:00                           NaN   
149  Person search  2025-05-31T22:31:38+00:00                           NaN   

     Policing operation   Latitude  Longitude Gender Age range  \
0                   NaN  51.517461  -0.089808   Male     18-24   
1                   NaN  51.513882  -0.095924   Male     18-24   
2                   NaN  51.512024  -0.094460   Male   over 34   
3                   NaN  51.511794  -0.101906   Male   over 34   
4                   NaN  51.513079  -0.094070   Male   over 34   
..                  ...        ...        ...    ...       ...   
145                 NaN  51.517296  -0.081902   Male     18-24   
146                 NaN  51.517296  -0.081902   Male     25-34   
147                 NaN  51.517296  -0.081902   Male     25-34   
148                 NaN  51.506419  -0.088195   Male   over 34   
149                 NaN  51.518440  -0.078755   Male     10-17   

                                Self-defined ethnicity  \
0                         Asian/Asian British - Indian   
1                      Other ethnic group - Not stated   
2                      Other ethnic group - Not stated   
3                   White - Any other White background   
4    White - English/Welsh/Scottish/Northern Irish/...   
..                                                 ...   
145                    Other ethnic group - Not stated   
146                    Other ethnic group - Not stated   
147    Black/African/Caribbean/Black British - African   
148                 White - Any other White background   
149                    Other ethnic group - Not stated   

    Officer-defined ethnicity  \
0                       Asian   
1                       Black   
2                       Other   
3                       White   
4                       White   
..                        ...   
145                     Black   
146                     Black   
147                     Black   
148                     White   
149                     Other   

                                           Legislation  \
0    Police and Criminal Evidence Act 1984 (section 1)   
1    Police and Criminal Evidence Act 1984 (section 1)   
2    Police and Criminal Evidence Act 1984 (section 1)   
3    Police and Criminal Evidence Act 1984 (section 1)   
4                Misuse of Drugs Act 1971 (section 23)   
..                                                 ...   
145              Misuse of Drugs Act 1971 (section 23)   
146  Police and Criminal Evidence Act 1984 (section 1)   
147  Police and Criminal Evidence Act 1984 (section 1)   
148  Police and Criminal Evidence Act 1984 (section 1)   
149  Police and Criminal Evidence Act 1984 (section 1)   

             Object of search                          Outcome  \
0    Article for use in theft     A no further action disposal   
1    Article for use in theft  Caution (simple or conditional)   
2    Article for use in theft             Community resolution   
3                Stolen goods                           Arrest   
4            Controlled drugs             Community resolution   
..                        ...                              ...   
145          Contro

In [33]:
data = pd.concat([data1, data2, data3, data4, data5], ignore_index=True)
display(data.head())
display(data.tail())

Type                       Date  \
0  Person and Vehicle search  2025-01-01T19:44:54+00:00   
1  Person and Vehicle search  2025-01-01T19:53:21+00:00   
2  Person and Vehicle search  2025-01-01T20:19:35+00:00   
3  Person and Vehicle search  2025-01-01T23:12:51+00:00   
4              Person search  2025-01-02T03:19:44+00:00   

   Part of a policing operation  Policing operation   Latitude  Longitude  \
0                           NaN                 NaN        NaN        NaN   
1                           NaN                 NaN        NaN        NaN   
2                           NaN                 NaN        NaN        NaN   
3                           NaN                 NaN        NaN        NaN   
4                           NaN                 NaN  51.510042  -0.120439   

  Gender Age range                           Self-defined ethnicity  \
0   Male     18-24                Asian/Asian British - Bangladeshi   
1   Male     18-24                Asian/Asian British - Bangladeshi   
2   Male   over 34                  Other ethnic group - Not stated   
3   Male     25-34  Black/African/Caribbean/Black British - African   
4   Male     18-24                  Other ethnic group - Not stated   

  Officer-defined ethnicity                            Legislation  \
0                     Asian  Misuse of Drugs Act 1971 (section 23)   
1                     White  Misuse of Drugs Act 1971 (section 23)   
2                     Asian  Misuse of Drugs Act 1971 (section 23)   
3                     Black  Misuse of Drugs Act 1971 (section 23)   
4                     Asian  Misuse of Drugs Act 1971 (section 23)   

   Object of search                       Outcome  \
0  Controlled drugs  A no further action disposal   
1  Controlled drugs  A no further action disposal   
2  Controlled drugs  A no further action disposal   
3  Controlled drugs  A no further action disposal   
4  Controlled drugs  A no further action disposal   

   Outcome linked to object of search  \
0                               False   
1                                True   
2                               False   
3                               False   
4                               False   

   Removal of more than just outer clothing  
0                                     False  
1                                     False  
2                                     False  
3                                     False  
4                                     False

Type                       Date  Part of a policing operation  \
720  Person search  2025-05-31T04:25:52+00:00                           NaN   
721  Person search  2025-05-31T05:38:03+00:00                           NaN   
722  Person search  2025-05-31T05:47:40+00:00                           NaN   
723  Person search  2025-05-31T12:51:55+00:00                           NaN   
724  Person search  2025-05-31T22:31:38+00:00                           NaN   

     Policing operation   Latitude  Longitude Gender Age range  \
720                 NaN  51.517296  -0.081902   Male     18-24   
721                 NaN  51.517296  -0.081902   Male     25-34   
722                 NaN  51.517296  -0.081902   Male     25-34   
723                 NaN  51.506419  -0.088195   Male   over 34   
724                 NaN  51.518440  -0.078755   Male     10-17   

                              Self-defined ethnicity  \
720                  Other ethnic group - Not stated   
721                  Other ethnic group - Not stated   
722  Black/African/Caribbean/Black British - African   
723               White - Any other White background   
724                  Other ethnic group - Not stated   

    Officer-defined ethnicity  \
720                     Black   
721                     Black   
722                     Black   
723                     White   
724                     Other   

                                           Legislation   Object of search  \
720              Misuse of Drugs Act 1971 (section 23)   Controlled drugs   
721  Police and Criminal Evidence Act 1984 (section 1)       Stolen goods   
722  Police and Criminal Evidence Act 1984 (section 1)       Stolen goods   
723  Police and Criminal Evidence Act 1984 (section 1)  Offensive weapons   
724  Police and Criminal Evidence Act 1984 (section 1)       Stolen goods   

                          Outcome  Outcome linked to object of search  \
720  A no further action disposal                               False   
721  A no further action disposal                               False   
722  A no further action disposal                               False   
723  A no further action disposal                               False   
724  A no further action disposal                               False   

     Removal of more than just outer clothing  
720                                     False  
721                                     False  
722                                     False  
723                                     False  
724                                     False

In [34]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 725 entries, 0 to 724
Data columns (total 15 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   Type                                      725 non-null    object 
 1   Date                                      725 non-null    object 
 2   Part of a policing operation              0 non-null      float64
 3   Policing operation                        0 non-null      float64
 4   Latitude                                  558 non-null    float64
 5   Longitude                                 558 non-null    float64
 6   Gender                                    725 non-null    object 
 7   Age range                                 661 non-null    object 
 8   Self-defined ethnicity                    724 non-null    object 
 9   Officer-defined ethnicity                 724 non-null    object 
 10  Legislation                           

In [35]:
data.isnull().any()


Type                                        False
Date                                        False
Part of a policing operation                 True
Policing operation                           True
Latitude                                     True
Longitude                                    True
Gender                                      False
Age range                                    True
Self-defined ethnicity                       True
Officer-defined ethnicity                    True
Legislation                                 False
Object of search                             True
Outcome                                     False
Outcome linked to object of search          False
Removal of more than just outer clothing    False
dtype: bool

เช็กว่าคอลัมน์ไหนมีค่า missing (NaN)

คืนค่าเป็น True/False สำหรับแต่ละคอลัมน์

จะเห็นว่าตัวแปร Object of search มี missing  อยากดูว่ามีกี่ record

In [36]:
print(data['Object of search'].isnull().sum())

4


In [37]:
dataupdate = data.dropna(subset=['Object of search'])
display(dataupdate)

Type                       Date  \
0    Person and Vehicle search  2025-01-01T19:44:54+00:00   
1    Person and Vehicle search  2025-01-01T19:53:21+00:00   
2    Person and Vehicle search  2025-01-01T20:19:35+00:00   
3    Person and Vehicle search  2025-01-01T23:12:51+00:00   
4                Person search  2025-01-02T03:19:44+00:00   
..                         ...                        ...   
720              Person search  2025-05-31T04:25:52+00:00   
721              Person search  2025-05-31T05:38:03+00:00   
722              Person search  2025-05-31T05:47:40+00:00   
723              Person search  2025-05-31T12:51:55+00:00   
724              Person search  2025-05-31T22:31:38+00:00   

     Part of a policing operation  Policing operation   Latitude  Longitude  \
0                             NaN                 NaN        NaN        NaN   
1                             NaN                 NaN        NaN        NaN   
2                             NaN                 NaN        NaN        NaN   
3                             NaN                 NaN        NaN        NaN   
4                             NaN                 NaN  51.510042  -0.120439   
..                            ...                 ...        ...        ...   
720                           NaN                 NaN  51.517296  -0.081902   
721                           NaN                 NaN  51.517296  -0.081902   
722                           NaN                 NaN  51.517296  -0.081902   
723                           NaN                 NaN  51.506419  -0.088195   
724                           NaN                 NaN  51.518440  -0.078755   

    Gender Age range                           Self-defined ethnicity  \
0     Male     18-24                Asian/Asian British - Bangladeshi   
1     Male     18-24                Asian/Asian British - Bangladeshi   
2     Male   over 34                  Other ethnic group - Not stated   
3     Male     25-34  Black/African/Caribbean/Black British - African   
4     Male     18-24                  Other ethnic group - Not stated   
..     ...       ...                                              ...   
720   Male     18-24                  Other ethnic group - Not stated   
721   Male     25-34                  Other ethnic group - Not stated   
722   Male     25-34  Black/African/Caribbean/Black British - African   
723   Male   over 34               White - Any other White background   
724   Male     10-17                  Other ethnic group - Not stated   

    Officer-defined ethnicity  \
0                       Asian   
1                       White   
2                       Asian   
3                       Black   
4                       Asian   
..                        ...   
720                     Black   
721                     Black   
722                     Black   
723                     White   
724                     Other   

                                           Legislation   Object of search  \
0                Misuse of Drugs Act 1971 (section 23)   Controlled drugs   
1                Misuse of Drugs Act 1971 (section 23)   Controlled drugs   
2                Misuse of Drugs Act 1971 (section 23)   Controlled drugs   
3                Misuse of Drugs Act 1971 (section 23)   Controlled drugs   
4                Misuse of Drugs Act 1971 (section 23)   Controlled drugs   
..                                                 ...                ...   
720              Misuse of Drugs Act 1971 (section 23)   Controlled drugs   
721  Police and Criminal Evidence Act 1984 (section 1)       Stolen goods   
722  Police and Criminal Evidence Act 1984 (section 1)       Stolen goods   
723  Police and Criminal Evidence Act 1984 (section 1)  Offensive weapons   
724  Police and Criminal Evidence Act 1984 (section 1)       Stolen goods   

                          Outcome  Outcome linked to object of search  \
0    A no further action disposal                               False   

In [60]:
dataupdate.isnull().any()


Type                                        False
Date                                        False
Part of a policing operation                 True
Policing operation                           True
Latitude                                     True
Longitude                                    True
Gender                                      False
Age range                                    True
Self-defined ethnicity                       True
Officer-defined ethnicity                    True
Legislation                                 False
Object of search                            False
Outcome                                     False
Outcome linked to object of search          False
Removal of more than just outer clothing    False
dtype: bool

In [61]:
dataupdate.groupby('Object of search')['Outcome'].apply(list) #รวม ให้สินค้ารหัสใบเสร็จเดียวกันมาอยู่ด้วยกัน # จัดกลุ่มข้อมูล .apply(func  ที่จะทำ)

Object of search
Article for use in theft               [Arrest, A no further action disposal, A no fu...
Articles for use in criminal damage    [A no further action disposal, Community resol...
Controlled drugs                       [A no further action disposal, A no further ac...
Evidence of offences under the Act     [A no further action disposal, A no further ac...
Firearms                               [A no further action disposal, A no further ac...
Offensive weapons                      [A no further action disposal, A no further ac...
Psychoactive substances                [A no further action disposal, A no further ac...
Stolen goods                           [A no further action disposal, A no further ac...
Name: Outcome, dtype: object

In [81]:
transactions = dataupdate.dropna(subset=['Object of search','Outcome']) \
    .apply(lambda r: [f"Object={r['Object of search']}",
                      f"Outcome={r['Outcome']}"], axis=1) \
    .tolist()


#กำหนดให้ Transaction = 1 เหตุการณ์การค้น (1 แถว)  
#Items = สิ่งที่ค้นหา (Object) + ผลลัพธ์ (Outcome) ของเหตุการณ์นั้น

In [82]:
# 1 record → ["Object=...", "Outcome=..."]
transactions = dataupdate.apply(
    lambda r: [f"Object={r['Object of search']}",
               f"Outcome={r['Outcome']}"],
    axis=1
).tolist()

In [56]:
!pip install apyori
from apyori import apriori

In [84]:
list(apriori(transactions, min_support=0.5))


[RelationRecord(items=frozenset({'Outcome=A no further action disposal'}), support=0.7045769764216366, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Outcome=A no further action disposal'}), confidence=0.7045769764216366, lift=1.0)])]

In [85]:
list(apriori(transactions, min_support=0.065))


[RelationRecord(items=frozenset({'Object=Article for use in theft'}), support=0.17198335644937587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Object=Article for use in theft'}), confidence=0.17198335644937587, lift=1.0)]),
 RelationRecord(items=frozenset({'Object=Controlled drugs'}), support=0.4826629680998613, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Object=Controlled drugs'}), confidence=0.4826629680998613, lift=1.0)]),
 RelationRecord(items=frozenset({'Object=Offensive weapons'}), support=0.10957004160887657, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Object=Offensive weapons'}), confidence=0.10957004160887657, lift=1.0)]),
 RelationRecord(items=frozenset({'Object=Stolen goods'}), support=0.17198335644937587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Object=Stolen goods'}), confidence=0.17198335644937587, lift=1.0)]),
 Rel

support=0.171983... (~17.2%)
หมายถึง itemset นี้ปรากฏใน 17.2% ของธุรกรรมทั้งหมด (17.2% ของเหตุการณ์ stop & search มีการค้นหา “Article for use in theft”).

In [86]:
frequentItemsets = list(apriori(transactions, min_support=0.065))
frequentItemsets[-1]

RelationRecord(items=frozenset({'Object=Stolen goods', 'Outcome=A no further action disposal'}), support=0.11789181692094314, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Object=Stolen goods', 'Outcome=A no further action disposal'}), confidence=0.11789181692094314, lift=1.0), OrderedStatistic(items_base=frozenset({'Object=Stolen goods'}), items_add=frozenset({'Outcome=A no further action disposal'}), confidence=0.6854838709677419, lift=0.9729013208026415), OrderedStatistic(items_base=frozenset({'Outcome=A no further action disposal'}), items_add=frozenset({'Object=Stolen goods'}), confidence=0.1673228346456693, lift=0.9729013208026416)])

frequentItemsets = list(apriori(transactions, min_support=0.065))
frequentItemsets[-1]
* คู่ {Stolen goods, NFA} เกิดร่วมกัน ~11.8%
* ในเคสที่ค้นหา “Stolen goods” ประมาณ 68.6% จบแบบ NFA


In [69]:
frequentItemsets[-1][0]

frozenset({'Object=Stolen goods', 'Outcome=A no further action disposal'})

ดู itemset สุดท้าย (-1 ) ประกอบด้วย 2 รายการ: ถูกค้นหา “Stolen goods” และ ผลลัพธ์ NFA เกิด พร้อมกัน

In [70]:
len(frequentItemsets[-1][0])


2

นับจำนวนสมาชิกใน itemset สุดท้ายได้ 2 ตัว

In [87]:
min_sup = 0.065
frequentItemsets = list(apriori(transactions, min_support=min_sup))

for i in frequentItemsets:
    if len(i[0]) > 1:   # ถ้ามีมากกว่า 1 item
        print("yes")
    print(i)


RelationRecord(items=frozenset({'Object=Article for use in theft'}), support=0.17198335644937587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Object=Article for use in theft'}), confidence=0.17198335644937587, lift=1.0)])
RelationRecord(items=frozenset({'Object=Controlled drugs'}), support=0.4826629680998613, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Object=Controlled drugs'}), confidence=0.4826629680998613, lift=1.0)])
RelationRecord(items=frozenset({'Object=Offensive weapons'}), support=0.10957004160887657, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Object=Offensive weapons'}), confidence=0.10957004160887657, lift=1.0)])
RelationRecord(items=frozenset({'Object=Stolen goods'}), support=0.17198335644937587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'Object=Stolen goods'}), confidence=0.17198335644937587, lift=1.0)])
RelationReco

support = 0.11096 (~11.1%)
~11.1% ของเหตุการณ์ทั้งหมดเป็นคู่ (Article for use in theft ∧ NFA)

กฎ 1 (ฐานว่าง → คู่) – ข้ามได้ (เชิงเทคนิค)

confidence = support, lift = 1.0

กฎ 2: {Object=Article for use in theft} → {Outcome=NFA}

confidence = 0.64516 (64.5%) = ถ้าเจอ Article for use in theft แล้ว 64.5% จบ NFA

lift = 0.91567 (< 1) = โอกาส NFA ต่ำกว่า*ค่าเฉลี่ยทั้งชุดเล็กน้อย
(baseline NFA ≈ 0.64516 / 0.91567 ≈ 70.5%)

กฎ 3: {Outcome=NFA} → {Object=Article for use in theft}

confidence = 0.15748 (15.7%) = ในเคสที่จบ NFA มี ~15.7% ที่เป็น Article for use in theft

In [72]:
def extract_rules(records):
    rows = []
    for r in records:
        for stat in r.ordered_statistics:
            A = list(stat.items_base)
            C = list(stat.items_add)
            if len(C)==1 and C[0].startswith('Outcome=') and any(x.startswith('Object=') for x in A):
                rows.append({
                    'antecedent': ', '.join(A),
                    'consequent': C[0],
                    'support': r.support,
                    'confidence': stat.confidence,
                    'lift': stat.lift
                })
    import pandas as pd
    return (pd.DataFrame(rows)
              .sort_values(['lift','confidence','support'], ascending=False)
              .reset_index(drop=True))

rules_df = extract_rules(frequentItemsets)
rules_df


antecedent                            consequent  \
0         Object=Offensive weapons  Outcome=A no further action disposal   
1          Object=Controlled drugs                        Outcome=Arrest   
2              Object=Stolen goods  Outcome=A no further action disposal   
3          Object=Controlled drugs  Outcome=A no further action disposal   
4  Object=Article for use in theft  Outcome=A no further action disposal   

    support  confidence      lift  
0  0.099861    0.911392  1.293531  
1  0.102635    0.212644  0.976536  
2  0.117892    0.685484  0.972901  
3  0.323162    0.669540  0.950273  
4  0.110957    0.645161  0.915672

In [73]:
print(dataupdate['Object of search'].unique())

['Controlled drugs' 'Stolen goods' 'Article for use in theft'
 'Offensive weapons' 'Evidence of offences under the Act' 'Firearms'
 'Articles for use in criminal damage' 'Psychoactive substances']


In [74]:
print(dataupdate['Outcome'].unique())

['A no further action disposal' 'Arrest' 'Community resolution'
 'Summons / charged by post' 'Khat or Cannabis warning'
 'Caution (simple or conditional)']
